From 31aa39262e0bfe7812695b79f07d3ebf7d8346ac Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Tue, 18 Nov 2014 20:25:15 -0500
Subject: [scsi] TUR path is down after adapter gets reset with multipath

Message-id: <1416342315-19881-3-git-send-email-emilne@redhat.com>
Patchwork-id: 100259
O-Subject: [RHEL7.1 PATCH e-stor 2/2] scsi: TUR path is down after adapter gets reset with multipath
Bugzilla: 1153738
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

upstream commit 333b2448cf5bd5a94c91a77136cac837d38fb984
Author: wenxiong@linux.vnet.ibm.com <wenxiong@linux.vnet.ibm.com>
Date:   Thu Nov 6 15:11:23 2014 -0600

    scsi: TUR path is down after adapter gets reset with multipath

    This patch fixes an issue with multipath ipr SAS devices which require a
    start unit command to be issued following an adapter reset. Without this
    patch, paths get marked failed following an adapter reset and since the
    error handler never gets invoked to issue the start unit, the paths are
    never recovered. Returning FAILED for this case ensures the error
    handler wakes up to issue the start unit.

    Signed-off-by: Brian King <brking@linux.vnet.ibm.com>
    Tested-by: Wen Xiong <wenxiong@linux.vnet.ibm.com>
    Reviewed-by: Hannes Reinecke <hare@suse.de>
    Signed-off-by: Christoph Hellwig <hch@lst.de>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/device_handler/scsi_dh_alua.c b/drivers/scsi/device_handler/scsi_dh_alua.c
index 7bcf67e..b8729ca 100644
--- a/drivers/scsi/device_handler/scsi_dh_alua.c
+++ b/drivers/scsi/device_handler/scsi_dh_alua.c
@@ -474,6 +474,13 @@ static int alua_check_sense(struct scsi_device *sdev,
     * LUN Not Ready -- Offline
     */
    return SUCCESS;
+  if (sdev->allow_restart &&
+      sense_hdr->asc == 0x04 && sense_hdr->ascq == 0x02)
+   /*
+    * if the device is not started, we need to wake
+    * the error handler to start the motor
+    */
+   return FAILED;
   break;
  case UNIT_ATTENTION:
   if (sense_hdr->asc == 0x29 && sense_hdr->ascq == 0x00)
-- 
1.7.1